cmd/tk/export.go

```diff
@@ -1,137 +1,87 @@
 
 	vars := workflowFlags(cmd.Flags())
 	getJsonnetOpts := jsonnetFlags(cmd.Flags())
 
 	cmd.Run = func(cmd *cli.Command, args []string) error {
-		// dir must be empty
-		to := args[1]
-		empty, err := dirEmpty(to)
-		if err != nil {
-			return fmt.Errorf("Checking target dir: %s", err)
-		}
-		if !empty && !*merge {
-			return fmt.Errorf("Output dir `%s` not empty. Pass --merge to ignore this", to)
-		}
-
-		// exit early if the template is bad
-
-		// Replace all os.path separators in string with BelRune for creating subfolders
-		replacedFormat := strings.Replace(*format, string(os.PathSeparator), BelRune, -1)
-
-		tmpl, err := template.New("").
-			Funcs(sprig.TxtFuncMap()). // register Masterminds/sprig
-			Parse(replacedFormat)      // parse template
-		if err != nil {
-			return fmt.Errorf("Parsing name format: %s", err)
-		}
-
-		// get the manifests
-		res, err := tanka.Show(args[0], tanka.Opts{
-			JsonnetOpts: getJsonnetOpts(),
-			Filters:     stringsToRegexps(vars.targets),
-		})
-		if err != nil {
-			return err
-		}
-
-		// write each to a file
-		for _, m := range res {
-			buf := bytes.Buffer{}
-			if err := tmpl.Execute(&buf, m); err != nil {
-				log.Fatalln("executing name template:", err)
-			}
-
-			// Replace all os.path separators in string in order to not accidentally create subfolders
-			name := strings.Replace(buf.String(), string(os.PathSeparator), "-", -1)
-			// Replace the BEL character inserted with a path separator again in order to create a subfolder
-			name = strings.Replace(name, BelRune, string(os.PathSeparator), -1)
-
-			// Create all subfolders in path
-			path := filepath.Join(to, name+"."+*extension)
-
-			// Abort if already exists
-			if exists, err := fileExists(path); err != nil {
+		var selector labels.Selector
+		var err error
+		if *labelSelector != "" {
+			selector, err = labels.Parse(*labelSelector)
+			if err != nil {
 				return err
-			} else if exists {
-				return fmt.Errorf("File '%s' already exists. Aborting", path)
 			}
+		}
 
-			// Write file
-			if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
-				return fmt.Errorf("creating filepath '%s': %s", filepath.Dir(path), err)
-			}
-			data := m.String()
-			if err := ioutil.WriteFile(path, []byte(data), 0644); err != nil {
-				return fmt.Errorf("writing manifest: %s", err)
+		var paths []string
+		for _, path := range args[1:] {
+			if *recursive {
+				envs, err := tanka.FindEnvironments(path, selector)
+				if err != nil {
+					return err
+				}
+				for _, env := range envs {
+					paths = append(paths, env.Metadata.Namespace)
+				}
+			} else {
+				opts := tanka.ParseOpts{
+					Evaluator: tanka.EnvsOnlyEvaluator,
+					Selector:  selector,
+				}
+				_, _, err := tanka.ParseEnv(path, opts)
+				if err != nil {
+					return err
+				}
+				paths = append(paths, path)
 			}
 		}
 
-		return nil
+		opts := tanka.ExportEnvOpts{
+			Format:    *format,
+			DirFormat: *dirFormat,
```

Comment:
What's the reasoning behind having two templates here? The BelRune logic introduces in an earlier PR allows `Format` to include slashes, which should be sufficient for arbitrarily complex directory layouts.

Unless I miss something, the above functionality seems to introduce competing ways to lay out directories.

Reply:
`Format` is being templated by the Kubernetes object, as it was before.

`DirFormat` is being templated by the tanka.dev/Environment object.

The latter allows us to define a directory structure based on metadata in the Environment object, for example `metadata.labels.cluster_name` could be useful but it would be a bit too verbose to have that label on every object.

Reply:
Ah, I see.

I am in favor of having just a single flag, for reasons stated above.

Wdyt about binding a function called "env" into the template, so you can do:

```
{env.metadata.name}/{.kind}-{.metadata.name}.yaml
```

This would combine the functionality into a single template